src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -651,12 +651,19 @@ void C2_MacroAssembler::fast_lock(Register objReg, Register boxReg, Register tmp
   movptr(scrReg, boxReg);
   movptr(boxReg, tmpReg);                   // consider: LEA box, [tmp-2]
 
-  xorptr(tmpReg, tmpReg);
+  // Optimistic form: consider XORL tmpReg,tmpReg
+  movptr(tmpReg, NULL_WORD);
 
-  // Appears unlocked - try to swing _owner from null to current thread.
+  // Appears unlocked - try to swing _owner from null to non-null.
+  // Ideally, I'd manifest "Self" with get_thread and then attempt
+  // to CAS the register containing Self into m->Owner.
+  // But we don't have enough registers, so instead we can either try to CAS
+  // rsp or the address of the box (in scr) into &m->owner.  If the CAS succeeds
+  // we later store "Self" into m->Owner.  Transiently storing a stack address
+  // (rsp or the address of the box) into  m->owner is harmless.
   // Invariant: tmpReg == 0.  tmpReg is EAX which is the implicit cmpxchg comparand.
   lock();
-  cmpxchgptr(thread, Address(boxReg, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner)));
+  cmpxchgptr(scrReg, Address(boxReg, OM_OFFSET_NO_MONITOR_VALUE_TAG(owner)));
```

Comment:
Sigh... I had liked the fact that we took care of these old "TODO" items
in this code. It's true that these changes were in violation of our "try not
to change stack-lock" mantra. I did run the v66 changes thru Mach5
Tier[1-8] testing in "stack-locking is default" mode so your changes
were well tested.

Reply:
Let's re-do those changes in a follow-up, ok?

Reply:
I've filed: https://bugs.openjdk.org/browse/JDK-8307493